src/Compilers/Core/Portable/Diagnostic/ExternalFileLocation.cs

```diff
@@ -14,12 +14,13 @@ namespace Microsoft.CodeAnalysis
     internal sealed class ExternalFileLocation : Location, IEquatable<ExternalFileLocation?>
     {
         private readonly TextSpan _sourceSpan;
-        private readonly FileLinePositionSpan _lineSpan;
+        private readonly FileLinePositionSpan _lineSpan, _mappedLineSpan;
 
-        internal ExternalFileLocation(string filePath, TextSpan sourceSpan, LinePositionSpan lineSpan)
+        internal ExternalFileLocation(string filePath, TextSpan sourceSpan, LinePositionSpan lineSpan, string mappedFilePath, LinePositionSpan mappedLineSpan)
         {
             _sourceSpan = sourceSpan;
             _lineSpan = new FileLinePositionSpan(filePath, lineSpan);
+            _mappedLineSpan = new FileLinePositionSpan(mappedFilePath, mappedLineSpan, hasMappedPath: !PathUtilities.PathsEqual(filePath, mappedFilePath));
```

Comment:
>!PathUtilities.PathsEqual(filePath, mappedFilePath) [](http://example.com/codeflow?start=102&length=51)

What if ```lineSpan``` != ```mappedLineSpan```
I do not think we should use any heuristic here. Consider having two separate constructors here. One that takes mapped information and one that doesn't. The one that takes mapped information always passes true here.